samples/core/guide/autograph.ipynb

```diff
@@ -609,6 +646,217 @@
+        "\n",
+        "\n",
+        "### Stateless functions\n",
+        "\n",
+        "For stateless functions like `collatz`, below, the easiest way to include them in a keras model is to wrap them up as a layer uisng `tf.keras.layers.Lambda`."
+      ]
+    },
+    {
+      "metadata": {
+        "id": "ChZh3q-zcF6C",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "cell_type": "code",
+      "source": [
+        "import numpy as np\n",
+        "\n",
+        "@autograph.convert()\n",
+        "def collatz(x):\n",
+        "  x=tf.reshape(x,())\n",
+        "  assert x>0\n",
+        "  n = tf.convert_to_tensor((0,)) \n",
+        "  while not tf.equal(x,1):\n",
+        "    n+=1\n",
+        "    if tf.equal(x%2, 0):\n",
+        "      x = x//2\n",
+        "    else:\n",
+        "      x = 3*x+1\n",
+        "      \n",
+        "  return n\n",
+        "\n",
+        "with tf.Graph().as_default():\n",
+        "  model = tf.keras.Sequential([\n",
+        "    tf.keras.layers.Lambda(collatz, input_shape=(1,), output_shape=(), )\n",
+        "  ])\n",
+        "  \n",
+        "result = model.predict(np.array([6171])) #261\n",
+        "result"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "metadata": {
+        "id": "k9LEoa3ud9hA",
+        "colab_type": "text"
+      },
+      "cell_type": "markdown",
+      "source": [
+        "### Custom Layers and Models\n",
+        "\n",
+        "<!--TODO(markdaoust) link to full examples  or these referenced models.-->\n",
+        "\n",
+        "The easiest way to use autograph is keras layers and models is to `@autograph.convert()` the `call` method. See the [keras guide](https://tensorflow.org/guide/keras#build_advanced_models) for details on how to build on these classes. \n",
+        "\n",
+        "Here is a simple example of the [stocastic network depth](https://arxiv.org/abs/1603.09382) technique :"
+      ]
+    },
+    {
+      "metadata": {
+        "id": "DJi_RJkeeOju",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "cell_type": "code",
+      "source": [
+        "# `K` is used to check if we're in train or test mode.\n",
+        "import tensorflow.keras.backend as K\n",
+        "\n",
+        "class StocasticNetworkDepth(tf.keras.Sequential):\n",
+        "  def __init__(self, pfirst=1.0, plast=0.5, *args,**kwargs):\n",
+        "    self.pfirst = pfirst\n",
+        "    self.plast = plast\n",
+        "    super().__init__(*args,**kwargs)\n",
+        "        \n",
+        "  def build(self,input_shape):\n",
+        "    super().build(input_shape.as_list())\n",
+        "    self.depth = len(self.layers)\n",
+        "    self.plims = np.linspace(self.pfirst, self.plast, self.depth+1)[:-1]\n",
+        "    \n",
+        "  @autograph.convert()\n",
+        "  def call(self, inputs):\n",
+        "    training = tf.cast(K.learning_phase(), dtype=bool)  \n",
```

Comment:
FYI, we're looking into modifying the default behavior to avoid the need for these explicit casts.

Reply:
+1, good to know.